FC-157 デモアプリとエミュレーター間でのメールアドレスを使用したアカウント認証処理作成

RikitoNoto opened the pull request:
# ✨ What's done

- [x] デモアプリとエミュレーター間でのメールアドレスを使用したアカウント認証処理

# 🚩 What's not done

- [ ] メールリンク認証

# ✅ Test

- [x] Androidエミュレーター、及びfirebase emulatorを使用してテスト
  - [x] firebase Authenticatioinに存在しているメールアドレスの場合、ホーム画面に遷移すること。
  - [x] firebase Authenticatioinに存在していないメールアドレスの場合、「メールアドレスが存在しません。」のエラーメッセージが出力されること。
  - [x] ホーム画面に遷移後、firebase Authenticatioinに登録されているUidが表示されること

# 🔧 補足、備考
## 目的
エミュレーターを使用したデモアプリ作成のため

# 🔀 マージ条件

- [ ] レビューを通過する
- [ ] セルフマージする

king-kazu39 committed:
FlutterFireの設定

king-kazu39 committed:
デモアプリにFirebaseの初期化処理を入れ忘れてたので追加

RikitoNoto committed:
[デモアプリ]firebaseの初期化処理作成

RikitoNoto committed:
[デモアプリ]ログイン処理の作成

RikitoNoto committed:
[デモページ]認証処理の作成/mainへの誤記削除

RikitoNoto committed:
[デモページ]firebaseへの依存をrepositoryのみになるように修正

RikitoNoto committed:
[デモページ]静的解析の内容を反映

king-kazu39 commented on experimental/pages/demologin/views/demologin_view.dart:
```diff
@@ -1,17 +1,18 @@
 class DemologinView extends GetView<DemologinController> {
   DemologinView({Key? key}) : super(key: key);
   final _formKey = GlobalKey<FormBuilderState>();
+  final TextEditingController _emailFieldController = TextEditingController();
```
```suggestion
```
`controller`を用意しないでもテキストフィールドの値取れるので不要になります。

RikitoNoto replied:
ご教示いただきありがとうございます！削除しました！

king-kazu39 commented on experimental/pages/demologin/views/demologin_view.dart:
```diff
@@ -31,6 +32,7 @@ class DemologinView extends GetView<DemologinController> {
                     padding: const EdgeInsets.symmetric(vertical: 20, horizontal: 30),
                     child: FormBuilderTextField(
                       name: _fieldNameEmail,
+                      controller: _emailFieldController,
```
```suggestion
```
不要なので削除お願いします

RikitoNoto replied:
ご教示いただきありがとうございます！削除しました！

king-kazu39 commented on experimental/pages/demologin/views/demologin_view.dart:
```diff
@@ -48,17 +50,21 @@ class DemologinView extends GetView<DemologinController> {
-                        //TODO: 認証の判定
-                        if(await Future.value(true)){
-                          Get.offNamed(DemoRoutes.demoHome);
+                        DemoUser? user = await controller.authentication(_emailFieldController.text);
```
```suggestion
                        DemoUser? user = await controller.authentication(_formKey.currentState!.value[_fieldNameEmail] ??
                              '');
```
これで値が取れるので変更お願いします

RikitoNoto replied:
ご教示いただきありがとうございます！
文字の取得方法を変更しました！

king-kazu39 commented on experimental/pages/demologin/views/demologin_view.dart:
```diff
@@ -48,17 +50,21 @@ class DemologinView extends GetView<DemologinController> {
-                          Get.offNamed(DemoRoutes.demoHome);
+                        DemoUser? user = await controller.authentication(_emailFieldController.text);
+                        // ユーザーを取得できた場合
+                        if(user != null){
+                          Get.offNamed(DemoRoutes.demoHome, arguments: user); // ホーム画面へ遷移
+                        }
+                        // ユーザーを取得できない場合
+                        else{
+                          _formKey.currentState?.invalidateField(name: _fieldNameEmail, errorText: "メールアドレスが存在しません。");
                         }
```
```suggestion
                          // ユーザーを取得できない場合
                          if (user == null) {
                            _formKey.currentState?.invalidateField(
                                name: _fieldNameEmail,
                                errorText: "メールアドレスが存在しません。");
                            return;
                          }
                          // ユーザーを取得できた場合
                          Get.offNamed(DemoRoutes.demoHome,
                              arguments: user); // ホーム画面へ遷移
                          return;
```
if else文多いので早期退出にしましょう

RikitoNoto replied:
修正しました！

king-kazu39 commented on experimental/repositories/demologin_repository.dart:
```diff
@@ -0,0 +1,22 @@
+
+  static Future<DemoUser?> authentication(String email) async {
+    try {
+      UserCredential credential = await FirebaseAuth.instance.signInWithEmailAndPassword(email: email, password: _password);
```
```suggestion
      UserCredential credential =
          await FirebaseAuth.instance.signInWithEmailAndPassword(
        email: email,
        password: _password,
      );
```
ちょっと長いので、コンマを入れて改行しましょう

RikitoNoto replied:
行を分割しました！

king-kazu39 commented on experimental/repositories/demologin_repository.dart:
```diff
@@ -0,0 +1,22 @@
+      }
+    }
+    catch (e) {
+      // nop
```
```suggestion
      // nop
      Get.log(e.toString);
```
log出力しときましょう

RikitoNoto replied:
修正しました！

king-kazu39 commented on experimental/pages/demologin/controllers/demologin_controller.dart:
```diff
@@ -1,9 +1,10 @@
 class DemologinController extends GetxController {
-  //TODO: Implement DemologinController
+  Future<DemoUser?> authentication(String email) async{
+    return DemologinRepository.authentication(email);
```
```suggestion
    return await DemologinRepository.authentication(email);
```
`await`つけましょう

RikitoNoto replied:
すみません、見落としていました。
追加しました！

RikitoNoto committed:
[デモページ]
- demologin_view.dart
 - コントローラを削除
 - ネストしたifをガード節に変更

- demologin_repository.dart
 - 長い1行を分割
 - 認証に失敗したときにログ出力するように変更

- demologin_controller.dart
 - awaitのつけ忘れを修正

king-kazu39 commented on experimental/pages/demologin/views/demologin_view.dart:
```diff
@@ -48,17 +48,20 @@ class DemologinView extends GetView<DemologinController> {
+                          return ;
                         }
+
+                        Get.offNamed(DemoRoutes.demoHome, arguments: user); // ホーム画面へ遷移
```
```suggestion
                        Get.offNamed(DemoRoutes.demoHome, arguments: user); // ホーム画面へ遷移
                        return;
```
`return`入れておいてほしいです。

RikitoNoto replied:
@king-kazu39 
かしこまりました！
入れる意味としては、この上にブロック節をパスした場合は
処理が`Get.offNamed(DemoRoutes.demoHome, arguments: user);`のみということを強調するためでしょうか？

この文の下のelse節にも入れたほうがいいのかわからなかったので、ご確認させていただきたいです。
(関数には必ず`return`をつけるルールとかでしょうか？)

king-kazu39 replied:
@RikitoNoto 
ご確認ありがとうございます！確かにそうですね！else節にも入れた方がいい気もしてきますね。returnを入れた方がここで処理が終わってると明示的に示せるので、私の好みですがどっちもreturnを入れちゃいますね。なので、Notoさんが今後もしここの処理を変更するときに迷わなければreturn入れなくてもいいですよ。

king-kazu39 commented on experimental/repositories/demologin_repository.dart:
```diff
@@ -0,0 +1,28 @@
+    try {
+      UserCredential credential = await FirebaseAuth.instance.signInWithEmailAndPassword(
+        email: email,
+        password: _password
```
```suggestion
        password: _password,
```
コンマ入れてないと保存した時に改行されないと思うのでコンマ入れておいてほしいです。

king-kazu39 commented:
@RikitoNoto 
修正対応ありがとうございます！
ちょっと見落としっぽいところがあったのでコメントしました。
修正したらプルリク出しちゃっていいですよ👍

RikitoNoto committed:
[デモページ]レビュー内容の反映

RikitoNoto commented:
> @RikitoNoto 修正対応ありがとうございます！ ちょっと見落としっぽいところがあったのでコメントしました。 修正したらプルリク出しちゃっていいですよ👍

ありがとうございます！
修正し、プルリク作成しましたのでご確認お願いします！
